openshift/dynamic/client.py

```diff
@@ -0,0 +1,370 @@
+#!/usr/bin/env python
+
+import sys
+import json
+from functools import partial
+
+import yaml
+from pprint import pformat
+
+from kubernetes import config
+from kubernetes.client.api_client import ApiClient
+from kubernetes.client.rest import ApiException
+
+
+class DynamicClient(object):
+
+    def __init__(self, client):
+        self.client = client
+        self.__resources = ResourceContainer(self.parse_api_groups())
+
+    @property
+    def resources(self):
+        return self.__resources
+
+    def default_groups(self):
+        groups = {}
+        groups['api'] = { '': {
+            'v1': self.get_resources_for_api_version('api', '', 'v1', True)
+        }}
+
+        try:
+            self.request('get', '/version/openshift')
+            groups['oapi'] = { '': {
+                'v1': self.get_resources_for_api_version('oapi', '', 'v1', True)
+            }}
+        except ApiException:
+            pass
+
+        return groups
+
+    def parse_api_groups(self):
+        ""
+        prefix = 'apis'
+        groups_response = self.request('GET', '/{}'.format(prefix))['groups']
+
+        groups = self.default_groups()
+        groups[prefix] = {}
+
+        for group in groups_response:
+            new_group = {}
+            for version_raw in group['versions']:
+                version = version_raw['version']
+                preferred = version_raw == group['preferredVersion']
+                new_group[version] = self.get_resources_for_api_version(prefix, group['name'], version, preferred)
+            groups[prefix][group['name']] = new_group
+        return groups
+
+    def get_resources_for_api_version(self, prefix, group, version, preferred):
+        """ returns a dictionary of resources associated with provided groupVersion"""
+
+        resources = {}
+
+        path = '/'.join(filter(None, [prefix, group, version]))
+        resources_response = self.request('GET', path)['resources']
+
+        # TODO: Filter out subresources for now
+        resources_raw = filter(lambda resource: '/' not in resource['name'], resources_response)
+
+        for resource in resources_raw:
+            resources[resource['kind']] = Resource(
+                prefix=prefix,
+                group=group,
+                api_version=version,
+                client=self,
+                preferred=preferred,
+                **resource
+            )
+        return resources
+
+    def get(self, resource, name=None, namespace=None, label_selector=None, field_selector=None):
+        path = resource.path(name=name, namespace=namespace)
+        return ResourceInstance(resource, self.request('get', path, label_selector=label_selector, field_selector=field_selector))
+
+    def create(self, resource, body, namespace=None):
+        if resource.namespaced:
+            namespace = namespace or body.get('metadata', {}).get('namespace')
```

Comment:
You can add a new function:

```
def _get_namespace(self, body, namespace=None):
    namespace = namespace or body.get('metadata', {}).get('namespace')
    if not namespace:
        raise Exception("namespace is required")
```
And update the 3 occurrences of that to:

```
namespace = self._get_namespace(body, namespace)
```

It feels worth it to me to avoid duplication unless necessary. You can let me know if it's gone overboard.

Reply:
On the same note, should you have a special exception class instead of raising the bare `Exception`?